### PR TITLE
[dbnode] Faster M3TSZ decoding by using 64 bit operations

### PR DESCRIPTION
**Pull request by linasm**
**What this PR does / why we need it**:
Changes `IStream` to read and process the encoded data stream by using 64 bit operations (instead of using the standard `io.Reader` that reads the data byte-by-byte).
According to the microbenchmark based on some real world time series, this **improves M3TSZ decoding performance by ~37%**. Microbenchmark results:

- before this PR -
`BenchmarkM3TSZDecode-12 10000 108797 ns/op`

- after switching to 64 bit decoding -
`BenchmarkM3TSZDecode-12 16813 71793 ns/op`

- after dropping IStream interface and using the struct directly -
`BenchmarkM3TSZDecode-12 16867 69272 ns/op`

Also included macro-benchmark results on some real world data, obtained using a [modified](https://github.com/m3db/m3/pull/2986) `read_data_files` utility:

Command: `read_data_files -b 1596009600000000000 -p ~/testdata/m3db -s 4 -v 0 -B datapoints`

Before the change:
```
Running time: 670.670802ms

7315 series read 
(10906.99 series/second)

3852592 datapoints decoded 
(5744386.05 datapoints/second)
```

After the change:
```
Running time: 460.262013ms

7315 series read 
(15893.12 series/second)

3852592 datapoints decoded 
(8370432.26 datapoints/second)
```
**31% improvement** (including the overhead of actually reading the file).

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
NONE

**Does this PR require updating code package or user-facing documentation?**:
NONE

### CODE REVIEW

**Comment by linasm** [comment]
Benchmark results:

- before this PR -
`BenchmarkM3TSZDecode-12    	   10000	    108797 ns/op`

- after switching to 64 bit decoding -
`BenchmarkM3TSZDecode-12    	   16813	      71793 ns/op`

- after dropping `IStream` interface and using the struct directly -
`BenchmarkM3TSZDecode-12    	   16867	      69272 ns/op`

File: src/dbnode/x/xio/segment_reader.go [hunk 1]
```diff
@@ -46,40 +47,101 @@ func (sr *segmentReader) Clone(
...
+		bytes byte
+	)
+
+	if sr.si >= nht {
+		return 0, 0, io.EOF
 	}
 
-	if b := sr.segment.Head; b != nil && len(sr.lazyHead) == 0 {
-		sr.lazyHead = b.Bytes()
+	if sr.si+8 < nh {
```
Comment (vpranckaitis) [inline src/dbnode/x/xio/segment_reader.go:64]: I believe this should be:

```suggestion
	if sr.si+8 <= nh {
```

Also a few more similar cases below.

File: src/dbnode/x/xio/segment_reader.go [hunk 2]
```diff
@@ -46,40 +47,101 @@ func (sr *segmentReader) Clone(
...
 	}
-	n := 0
-	if sr.si < nh {
-		nRead := copy(b, sr.lazyHead[sr.si:])
-		sr.si += nRead
-		n += nRead
-		if n == len(b) {
-			return n, nil
+
+	for ; sr.si < nht && bytes < 8; sr.si++ {
```
Comment (vpranckaitis) [inline src/dbnode/x/xio/segment_reader.go:92]: I think `bytes < 8` can be dropped (just like in [the `for` on `lazyHead`](https://github.com/m3db/m3/pull/2827/files#diff-3e6c8bee4811cf296ffb0bb8567b3272a641d696b0cc348e4a2a2984e6fa6c08R73-R76)), since otherwise previous `if` should have matched.

```suggestion
	for ; sr.si < nht; sr.si++ {
```

File: src/dbnode/x/xio/segment_reader_test.go [hunk 3]
```diff
@@ -48,6 +49,8 @@ var (
 		0xc0, 0x1, 0xf4, 0x1, 0x0, 0x0, 0x0, 0x2, 0x1, 0x2, 0x7, 0x10, 0x1e,
 		0x0, 0x1, 0x0, 0xe0, 0x65, 0x58, 0xcd, 0x3, 0x0, 0x0, 0x0, 0x0,
 	}
+
+	checkdNoPool = func(d []byte) checked.Bytes { return checked.NewBytes(d, nil) }
```
Comment (vpranckaitis) [inline src/dbnode/x/xio/segment_reader_test.go:53]:
```suggestion
	checkedNoPool = func(d []byte) checked.Bytes { return checked.NewBytes(d, nil) }
```

**Review by vpranckaitis — COMMENTED** [review]
LGTM with a few small comments

File: src/dbnode/encoding/istream_test.go [hunk 4]
```diff
@@ -107,14 +126,62 @@ func TestReadAfterPeekBits(t *testing.T) {
...
+
+func TestIStreamRemainingBitsInCurrentByte(t *testing.T) {
+	byteStream := []byte{0xff, 0, 0x42}
+	is := NewIStream(xio.NewBytesReader64(byteStream))
+	for _, b := range byteStream {
+		for i := 0; i < 8; i++ {
+			var expected uint
+			if i > 0 {
+				expected = uint(8 - i)
+			}
```
[the remark this answers is not in the document]
Comment (linasm) [reply src/dbnode/encoding/istream_test.go:168]: Yeah, since this is test code, I lean to keep the more clear version of it.

File: src/dbnode/x/xio/reader64_test.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+func TestBytesReader64(t *testing.T) {
+	var (
+		data = []byte{4, 5, 6, 7, 8, 9, 1, 2, 3, 0, 10, 11, 12, 13, 14, 15, 16, 17}
+		r    = NewBytesReader64(nil)
+	)
+
+	for l := 0; l < len(data); l++ {
+		testBytesReader64(t, r, data)
+	}
```
[the remark this answers is not in the document]
Comment (linasm) [reply src/dbnode/x/xio/reader64_test.go:40]: Good catch, it had to be `data[:l]`. Thankfully, the tests still pass.

File: src/dbnode/x/xio/segment_reader.go (hunk 1 above)
Comment (linasm) [reply src/dbnode/x/xio/segment_reader.go:64]: Good catch, will save a few cycles.

File: src/dbnode/x/xio/segment_reader.go (hunk 2 above)
Comment (linasm) [reply src/dbnode/x/xio/segment_reader.go:92]: Makes sense, fixing.

File: src/dbnode/x/xio/segment_reader_test.go (hunk 3 above)
Comment (linasm) [reply src/dbnode/x/xio/segment_reader_test.go:53]: 👍

**Comment by linasm** [comment]
Macro-benchmark results on some real world data, obtained using a modified `read_data_files` utility:

Command: `read_data_files -b 1596009600000000000 -p ~/testdata/m3db -s 4 -v 0 -B datapoints`

Before the change:
```
Running time: 670.670802ms

7315 series read
(10906.99 series/second)

3852592 datapoints decoded
(5744386.05 datapoints/second)
```

After the change:
```
Running time: 460.262013ms

7315 series read
(15893.12 series/second)

3852592 datapoints decoded
(8370432.26 datapoints/second)
```

31% improvement (including the overhead of actually reading the file).

**Comment by linasm** [comment]
Screenshot 2021-01-07 at 15 48 29
@robskillington we've noticed (in the CPU profile above) that M3TSZ decoding can take a significant chunk of peer bootstrappers processing time (decoding data streamed from peers). Which makes me think we should prioritise this PR already.

File: src/dbnode/encoding/istream.go [hunk 6]
```diff
@@ -21,169 +21,119 @@
...
-			numToRead = is.remaining
-		}
-		bits := is.current >> (8 - numToRead)
-		is.current <<= numToRead
-		is.remaining -= numToRead
-		res = (res << uint64(numToRead)) | uint64(bits)
-		numBits -= numToRead
+	res := readBitsInWord(is.current, numBits)
+	bitsNeeded := numBits - is.remaining
+	if err := is.readWordFromStream(); err != nil {
```
Comment (notbdu) [inline src/dbnode/encoding/istream.go:78]: Not following this logic here... shouldn't we check if there are remaining bits to be reading before reading another word from the stream? 

Might be worth adding some comments here around the logic.

File: src/dbnode/x/xio/reader64.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+// Read64 reads and returns a 64 bit word plus a number of bytes (up to 8) actually read.
+func (r *BytesReader64) Read64() (word uint64, n byte, err error) {
+	if r.index+8 <= len(r.data) {
+		// NB: this compiles to a single 64 bit load followed by
+		// a BSWAPQ on amd64 gc 1.13 (https://godbolt.org/z/oTK1jx).
+		res := binary.BigEndian.Uint64(r.data[r.index:])
+		r.index += 8
+		return res, 8, nil
+	}
+	if r.index >= len(r.data) {
```
Comment (notbdu) [inline src/dbnode/x/xio/reader64.go:48]: nit: maybe this check can go first? nbd though haha

File: src/dbnode/x/xio/segment_reader.go [hunk 8]
```diff
@@ -46,40 +47,101 @@ func (sr *segmentReader) Clone(
...
 }
 
-func (sr *segmentReader) Read(b []byte) (int, error) {
-	if len(b) == 0 {
-		return 0, nil
+func (sr *segmentReader) Read64() (word uint64, n byte, err error) {
+	sr.lazyInit()
+
+	var (
+		nh    = len(sr.lazyHead)
```
Comment (notbdu) [inline src/dbnode/x/xio/segment_reader.go:54]: nit: the variable names are a little hard to parse here

File: src/dbnode/encoding/istream.go [hunk 9]
```diff
@@ -21,169 +21,119 @@
...
-		return 0, is.err
-	}
-	if is.remaining == 0 {
-		if err := is.readByteFromStream(); err != nil {
-			return 0, err
-		}
-	}
-	return Bit(is.consumeBuffer(1)), nil
+// NewIStream creates a new IStream
+func NewIStream(reader64 xio.Reader64) *IStream {
```
[the remark this answers is not in the document]
Comment (linasm) [reply src/dbnode/encoding/istream.go:39]: A good question. `IStream` used to be an interface before this change, and now it has become a struct. I think I left it returning a pointer in order to avoid accidental semantics change at the call sites (to avoid any potential pass-by-value of this struct down the stream).

File: src/dbnode/encoding/istream.go (hunk 6 above)
Comment (linasm) [reply src/dbnode/encoding/istream.go:78]: There is such check with an early return few lines above:
```go
	if numBits <= is.remaining {
		// Have enough bits buffered.
		return is.consumeBuffer(numBits), nil
	}
```
I'll add some comments for clarity.

File: src/dbnode/x/xio/reader64.go (hunk 7 above)
Comment (linasm) [reply src/dbnode/x/xio/reader64.go:48]: Actually, I think the current implementation should be more performant. On most of the calls to this method we will have   8 (or more) bytes available, so we will do an early return from the first if, avoiding any other checks.

File: src/dbnode/x/xio/segment_reader.go (hunk 8 above)
Comment (linasm) [reply src/dbnode/x/xio/segment_reader.go:54]: Good point, will rename:
```go
		headLen     = len(sr.lazyHead)
		headTailLen = headLen + len(sr.lazyTail)
```

**Comment by notbdu** [comment]
Code semantically LGTM. Taking the perf measurements at face value :).

**Review by notbdu — APPROVED** [review]

**Review by vpranckaitis — APPROVED** [review]
